File: 0001-two-sum/0001-two-sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        # implement a HashMap or some kind of dictionary to hold key, value paris.
        pair_d = {}

        # Loop to iterate through the array
        for i,num in enumerate(nums):
            # conditional check to see if the number exists inside the dictionary.
            if num in pair_d: return pair_d[num], i
            # if it doesn't exist, sub the target value from num to get the second value and add it to the dictionary.
            pair_d[target-num]=i
            

File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
//     Understand the problem(DONE)
// Figure out a solution 
// BRUTE FORCE: check EVERY combination of the target for each value in the list.
//   Solution: use a Hashmap to store values as keys, and indices as values.
//  Iterate through the array and check if the difference between the target value and the array value sums up to the target.  So if the target is 4, and 3 is the first value in the array. 4-3 = 1.  3+1 = 4. Yes, add that to the Hashmap.
    
//  
    
// STEP ONE: create a Hashmap with  val: indice 
    let prevMap = {} 
    
// STEP TWO: iterate through nums
  for(let i = 0; i < nums.length; i++){
//       define value  and diff
       let n = nums[i];
      let diff = target - n;
      
      
//  checking if values exist in prevMap     
if (prevMap.hasOwnProperty(diff)) {
            // Return the indices when a solution is found
            return [prevMap[diff], i];
        }
//         add the value inside the Map if it does not exist 
        prevMap[n] = i;      
      

  }
    
 // Return null if no solution is found
    return null;         
}; 